core/vm.cpp

```diff
@@ -1308,6 +1312,77 @@ class Interpreter {
         return nullptr;
     }
 
+    const AST *builtinParseJson(const LocationRange &loc, const std::vector<Value> &args)
+    {
+        validateBuiltinArgs(loc, "parseJson", args, {Value::STRING});
+
+        std::string value = encode_utf8(static_cast<HeapString *>(args[0].v.h)->value);
+
+        auto j = json::parse(value);
+
+        bool filled;
+
+        otherJsonToHeap(j, filled, scratch);
+
+        return nullptr;
+    }
+
+    void otherJsonToHeap(const json &v, bool &filled, Value &attach) {
+        // In order to not anger the garbage collector, assign to attach immediately after
+        // making the heap object.
+        switch (v.type()) {
+            case json::value_t::string:
+                attach = makeString(decode_utf8(v.get<std::string>()));
+                filled = true;
+                break;
+
+            case json::value_t::boolean:
+                attach = makeBoolean(v.get<bool>());
+                filled = true;
+                break;
+
+            case json::value_t::number_integer:
+            case json::value_t::number_unsigned:
+            case json::value_t::number_float:
+                attach = makeNumber(v.get<double>());
+                filled = true;
+                break;
+
+            case json::value_t::null:
+                attach = makeNull();
+                filled = true;
+                break;
+
+            case json::value_t::array:{
+                attach = makeArray(std::vector<HeapThunk *>{});
+                filled = true;
+                auto *arr = static_cast<HeapArray *>(attach.v.h);
+                for (size_t i = 0; i < v.size(); ++i) {
+                    arr->elements.push_back(
+                        makeHeap<HeapThunk>(idArrayElement, nullptr, 0, nullptr));
+                    otherJsonToHeap(v[i], arr->elements[i]->filled, arr->elements[i]->content);
+                }
+            } break;
+
+            case json::value_t::object: {
+                attach = makeObject<HeapComprehensionObject>(
+                    BindingFrame{}, jsonObjVar, idJsonObjVar, BindingFrame{});
+                filled = true;
+                auto *obj = static_cast<HeapComprehensionObject *>(attach.v.h);
```

Comment:
Can't you make a regular simple object here?

Reply:
I don't see any reason why not. I used `HeapComprehensionObject`, because it is used here: https://github.com/google/jsonnet/blob/v0.11.2/core/vm.cpp#L1347. Do you remember why `HeapComprehensionObject` was used there?

Reply:
Oh it's because with HeapSimpleObject you have to make AST objects and those don't get garbage collected until the interpreter exits.  So using HeapComprehensionObject is basically a hack to minimize that (it's still leaky because it needs to make ids and those are not GC'd either).

Reply:
so we're leaving `HeapComprehensionObject`, I guess